### DISCUSSION TITLE
aioble vs RGB LED shared hardware resources

### DISCUSSION THREAD

**Discussion by lievenleroy** [micropython/micropython#13295] [RP2040 / Pico]
Another aioble issue on Nano RP2040 Connect:
The board's RGB pixel is connected to the wifi/Bluetooth chip. So setting it requires talking over SPI to that chip.

I have this working using sample code from [here](https://forum.arduino.cc/t/blinking-the-rgb-lights-in-micropython/868974). But once I initialize aioble, I can no longer reliably configure the pixel.

My BLE code is closely modeled on the [temperature server example](https://github.com/micropython/micropython-lib/blob/master/micropython/bluetooth/aioble/examples/temp_sensor.py) in the aioble Github. Roughly, my code looks like:
- init SPI interface for pixel
- init aioble
- start asyncio tasks for BLE advertising and to blink the pixel

Either task running solo works fine, but if I run both tasks the pixel fails to blink. I'm guessing aioble service interrupts are stepping on other SPI communication in progress. Is there a way to suspend aioble activity for a period of time (critical section)? Or a mutex or other means that the resource should be shared through?

**Comment by lievenleroy** [comment]
Updates:
- There's an alternate way to control the RBG LEDs via LED_BLUE/LED_GREEN/LED_RED pin definitions in the **machine** lib, but it has the same issue as when I control them using the code sample I mentioned above: doesn't work when aioble is in use
- Using a simple critical section (disable_irq/enable_irq) doesn't work, since it disables the ability to receive SPI responses

**Reply by jimmo** [reply]
@lievenleroy It would be interesting to know whether it's aioble or just bluetooth at all. i.e. if you just use the lower-level BLE api, e.g. https://github.com/micropython/micropython/blob/master/examples/bluetooth/ble_temperature.py

I'm not surprised that there are issues controlling the LEDs via SPI directly from Python, you should definitely use the `machine.Pin` abstraction here. But yes this should work.

Maybe @iabdalkader has some suggestions?

**Reply by iabdalkader** [reply]
@jimmo

Ah yes I saw the same issue a few days ago, I don't think it has anything to do with aioble. I think it's the firmware, it doesn't initialize the SPI peripheral/control pins when the module is reset in Bluetooth mode, and the SPI bus is required to control external pins. I'd suggest using the `LED` pin instead, which is controlled directly by the RP2040, while using Bluetooth. 

EDIT: Some more info, this function initializes bluetooth: https://github.com/arduino/nina-fw/blob/master/main/sketch.ino.cpp#L117
This one initializes WiFi (and SPI) : https://github.com/arduino/nina-fw/blob/master/main/sketch.ino.cpp#L159
You can see the firmware just takes a different path on reset depending on the reset mode. Note I'm not sure but there might be some shared pins between the BT/UART and the SPI interfaces too, need to check to confirm, so it might be impossible to use the two interfaces at the same time.

> I have this working using sample code from [here](https://forum.arduino.cc/t/blinking-the-rgb-lights-in-micropython/868974). But once I initialize aioble, I can no longer reliably configure the pixel.

@lievenleroy You don't need all that code to control the RGB LED, rp2's `machine.Pin` supports externally controlled GPIO, so you can just use `machine.Pin("LED").value(1)` and the same goes for externally controlled ADC channels btw (`machine.ADC` supports them).
